[CodeComplete] Fix cyclic dependency error in request evaluator during code completion

ahoppen opened the pull request:
[CodeComplete] Fix cyclic dependency error during code completion

In the added test case, we were hitting a cyclic dependency error in the request evaluator during code completion, that’s caued as follows:
– To complete at `#^COMPLETE^#`, we need to get the `NamingPattern` of `start`.
– Retrieving the `NamingPattern` of `start` causes the entire `if`-condition to be type checked, including `end`
– Type checking `end` requires getting the `NamingPattern` of `start`
=> Cyclic dependency

To resolve the issue, I added a special case to `NamingPatternRequest` that only type-checks the `StmtConditionElement` that actually defines the `VarDecl`.

Fixes rdar://75200217 [SR-14317]

ahoppen commented:
@swift-ci Please test

rintaro commented on lib/Sema/TypeChecker.h:
```diff
@@ -423,6 +423,12 @@ bool typesSatisfyConstraint(Type t1, Type t2, bool openArchetypes,
+/// Sets \p isFalsable to \c true if the condition might evaluate to \c false,
+/// otherwise leaves \p isFalsable untouched.
+/// \returns \c true if there was an error type checking, \c false otherwise.
+bool typeCheckStmtConditionElement(StmtConditionElement &elt, bool &isFalsable, DeclContext *dc);
```
80 columns

ahoppen replied:
Thanks. Forgot to run `git-clang-format` after finishing my changes.

rintaro commented on lib/Sema/TypeCheckStmt.cpp:
```diff
@@ -419,6 +419,87 @@ static LabeledStmt *findBreakOrContinueStmtTarget(
   return nullptr;
 }
 
+bool TypeChecker::typeCheckStmtConditionElement(StmtConditionElement &elt, bool &isFalsable, DeclContext *dc) {
```
80 columns

rintaro reviewed (commented):
It's nice that we don't type check unrelated conditions anymore :) But I'm a little worried that this PR changes the behavior of `getParenPatternStmt()`, especially `getRecursiveParentPatternStmt()`.

I think finding the element from `getParenPatternStmt()` is not so heavy.
```cpp
      if (auto *condStmt = dyn_cast<LabeledConditionalStmt>(parentStmt)) {
        for (auto &condElem: condStmt->getCond()) {
          if (auto *pat = condElem.getPatternOrNull()) {
            if (pat->containsVarDecl(VD)) {
              // Type check 'condElem'
              break;
            }
          }
        }
      }
```
Could we do this way without modifying `VarDecl`?

ahoppen commented:
> Could we do this way without modifying VarDecl?

I wanted to avoid creating a special case for `LabelConditionalStmt` in `NamingPatternRequest` because the basic idea that we need to type check the parent (whatever that may be) seemed right to me – I also wasn’t sure if there were other places that might benefit from only type checking one condition. But in hindsight, it would probably be better to just go for the special case, especially now that I know it’s only one place.

ahoppen commented:
Really good suggestion @rintaro. Simplified the change a lot!

ahoppen commented:
@swift-ci Please test

rintaro commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -2504,7 +2504,25 @@ NamingPatternRequest::evaluate(Evaluator &evaluator, VarDecl *VD) const {
+            // conditions
+            break;
+          }
+        }
```
Don't we need this?
```
namingPattern = VD->getCanonicalVarDecl()->NamingPattern;
```
I was thinking something like:
```cpp
if (auto parentStmt = VD->getParentPatternStmt()) {
  if (auto condStmt = dyn_cast_or_null<LabeledConditionalStmt>(parentStmt)) {
     // Only type check the condition we care about
     for (auto &cond : condStmt->getCond()) {
        ....
     }
  } else {
    // We have some other parent stmt. Type check it completely.
    if (auto CS = dyn_cast<CaseStmt>(parentStmt))
      parentStmt = CS->getParentStmt();
    ASTNode node(parentStmt);
    TypeChecker::typeCheckASTNode(node, VD->getDeclContext()
  }
   
  namingPattern = VD->getCanonicalVarDecl()->NamingPattern;
}
```

ahoppen replied:
I thought I could do two things simultaneously. Reality proves me wrong again 😬.

ahoppen committed:
[CodeComplete] Fix cyclic dependency error during code completion

In the added test case, we were hitting a cyclic dependency error in the request evaluator during code completion, that’s caued as follows:
– To complete at `#^COMPLETE^#`, we need to get the `NamingPattern` of `start`.
– Retrieving the `NamingPattern` of `start` causes the entire `if`-condition to be type checked, including `end`
– Type checking `end` requires getting the `NamingPattern` of `start`
=> Cyclic dependency

To resolve the issue, I added a special case to `NamingPatternRequest` that only type-checks the `StmtConditionElement` that actually defines the `VarDecl`.

Fixes rdar://75200217

ahoppen commented:
@swift-ci Please test

ahoppen commented:
@swift-ci Please test Windows
